mat/tridiag.go

```diff
@@ -0,0 +1,371 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+	"gonum.org/v1/gonum/lapack/lapack64"
+)
+
+var (
+	tridiagDense *Tridiag
+	_            Matrix           = tridiagDense
+	_            allMatrix        = tridiagDense
+	_            denseMatrix      = tridiagDense
+	_            Banded           = tridiagDense
+	_            MutableBanded    = tridiagDense
+	_            RawTridiagonaler = tridiagDense
+)
+
+// A RawTridiagonaler can return a lapack64.Tridiagonal representation of the
+// receiver. Changes to the DL, D, DU slices in lapack64.Tridiagonal will be
+// reflected in the original matrix, changes to the N field will not.
+type RawTridiagonaler interface {
+	RawTridiagonal() lapack64.Tridiagonal
+}
+
+// Tridiag represents a tridiagonal matrix by its three diagonals.
+type Tridiag struct {
+	mat lapack64.Tridiagonal
+}
+
+// NewTridiag creates a new n×n tridiagonal matrix with the first sub-diagonal
+// in dl, the main diagonal in d and the first super-diagonal in du. If dl, d,
+// and du are nil, new backing slices will be allocated for them. If dl and du
+// have length n-1 (or are nil if n is 1) and d has length n, they will be used
+// as backing slices, and changes to the elements of the returned Tridiag will
+// be reflected in dl, d, du. If neither of these is true, NewTridiag will
+// panic.
+func NewTridiag(n int, dl, d, du []float64) *Tridiag {
+	switch {
+	case n <= 0:
+		if n == 0 {
+			panic(ErrZeroLength)
+		}
+		panic(ErrNegativeDimension)
+	case n == 1:
+		if dl != nil || du != nil || (d != nil && len(d) != 1) {
+			panic(ErrShape)
+		}
```

Comment:
Can we allow `len(dl) == 0` and `len(du) == 0`?

Reply:
I think we can, I'll make the change.

Reply:
Done.